src/program.cpp

```diff
@@ -319,6 +324,8 @@ shape program::get_parameter_shape(std::string name) const
         return {};
 }
 
+std::vector<std::string> program::get_parameter_names() const { return impl->input_names; }
```

Comment:
This can return names to parameters that have been removed. This should iterate over the program and find all the parameters.

Reply:
We need to return the names of the same order as that in the onnx file. After compilation, we lose track of the name order in the program.

Reply:
Yea, there are several ways to do that. You can take the list of parameters found in the program and reverse them:

```cpp
std::vector<std::string> program::get_parameter_names() const 
{
    std::vector<std::string> result;
    for(auto&& ins : impl->instructions)
    {
        if(ins.name() == "@param")
        {
            auto&& name = any_cast<builtin::param>(ins.get_operator()).parameter;
            result.push_back(name);
        }
    }
    std::reverse(result.begin(), result.end();)
    return result;
}
```

Or check if `impl->input_names` is still a parameter:

```cpp
std::vector<std::string> program::get_parameter_names() const 
{
    std::vector<std::string> result = impl->input_names;
    std::unordered_set<std::string> params;
    for(auto&& ins : impl->instructions)
    {
        if(ins.name() == "@param")
        {
            auto&& name = any_cast<builtin::param>(ins.get_operator()).parameter;
            params.push_back(name);
        }
    }
    erase_if(result, [&](auto&& name) {
        return params.count(name) == 0;
    });
    return result;
}
```

Or add a field `order` to the parameter operator and sort them based on this:

```cpp
instruction_ref program::add_parameter(std::string name, shape s)
{
    assert(get_parameter_shape(name) == shape{});
    impl->instructions.push_front({builtin::param{std::move(name), nparam}, std::move(s), {}});
    nparam++;
    return impl->instructions.begin();
}
std::vector<std::string> program::get_parameter_names() const 
{
    std::vector<std::string> result;
    std::vector<builtin::param> params;
    for(auto&& ins : impl->instructions)
    {
        if(ins.name() == "@param")
        {
            auto&& param = any_cast<builtin::param>(ins.get_operator());
            params.push_back(param);
        }
    }
    std::sort(params.begin(), params.end(), by(std::less<>{}, [](auto&& p) { return p.order; }));
    std::transform(params.begin(), params.end(), std::back_inserter(params), [&](auto&& p) {
        return p.parameter;
    });
    return result;
}
```

Reply:
Thanks for the detailed info. I understand what you mean now. I think the second and third approaches work, so I just take the second one. For the first approach, the final name can be either the same or reversed of the order in the onnx file, so just reverse it may not work.